File: modern-digital-banking-dashboard/backend/app/services/reward_service.py
from typing import Dict, List, Optional, Tuple
from datetime import datetime, timedelta
from decimal import Decimal
import math
import logging

from app.schemas.reward import RewardTier
from app.core.config import settings

logger = logging.getLogger(__name__)

class RewardService:
    def __init__(self):
        # Define reward tiers with point thresholds
        self.tiers = {
            RewardTier.BRONZE: {
                "min_points": 0,
                "max_points": 499,
                "multiplier": 1.0,
                "benefits": ["Basic tracking", "Email support"],
                "color": "#cd7f32"  # Bronze color
            },
            RewardTier.SILVER: {
                "min_points": 500,
                "max_points": 1999,
                "multiplier": 1.1,
                "benefits": ["Priority support", "Advanced analytics", "Custom categories"],
                "color": "#c0c0c0"  # Silver color
            },
            RewardTier.GOLD: {
                "min_points": 2000,
                "max_points": 4999,
                "multiplier": 1.25,
                "benefits": ["All Silver benefits", "Early access to features", "Dedicated account manager"],
                "color": "#ffd700"  # Gold color
            },
            RewardTier.PLATINUM: {
                "min_points": 5000,
                "max_points": 9999,
                "multiplier": 1.5,
                "benefits": ["All Gold benefits", "Custom integrations", "API access", "White-label reports"],
                "color": "#e5e4e2"  # Platinum color
            },
            RewardTier.DIAMOND: {
                "min_points": 10000,
                "max_points": None,  # No upper limit
                "multiplier": 2.0,
                "benefits": ["All Platinum benefits", "24/7 phone support", "Custom development", "Enterprise features"],
                "color": "#b9f2ff"  # Diamond color
            }
        }
        
        # Define category multipliers
        self.category_multipliers = {
            "utilities": 1.0,
            "rent": 1.2,
            "mortgage": 1.2,
            "credit_card": 1.5,
            "loan": 1.3,
            "insurance": 1.1,
            "subscription": 0.8,
            "education": 1.4,
            "medical": 1.0,
            "tax": 1.0,
            "other": 1.0
        }
        
        # Base points per dollar
        self.base_points_per_dollar = 10
        
        # On-time payment bonus multiplier
        self.on_time_multiplier = 1.5
        
        # Streak bonus (consecutive on-time payments)
        self.streak_bonus = {
            3: 1.1,   # 10% bonus for 3+ streak
            7: 1.2,   # 20% bonus for 7+ streak
            15: 1.3,  # 30% bonus for 15+ streak
            30: 1.5   # 50% bonus for 30+ streak
        }
    
    def calculate_points(
        self, 
        bill_amount: Decimal, 
        on_time_payment: bool = True, 
        category: str = "other",
        streak_days: int = 0
    ) -> int:
        """
        Calculate reward points for a bill payment
        
        Args:
            bill_amount: Bill amount in USD
            on_time_payment: Whether payment was made on time
            category: Bill category
            streak_days: Consecutive days of on-time payments
        
        Returns:
            Calculated points (integer)
        """
        try:
            # Convert Decimal to float for calculation
            amount = float(bill_amount)
            
            # Base points based on amount
            base_points = amount * self.base_points_per_dollar
            
            # Apply category multiplier
            category_multiplier = self.category_multipliers.get(category.lower(), 1.0)
            points = base_points * category_multiplier
            
            # Apply on-time payment bonus
            if on_time_payment:
                points *= self.on_time_multiplier
            
            # Apply streak bonus if applicable
            streak_multiplier = 1.0
            for streak, multiplier in sorted(self.streak_bonus.items(), reverse=True):
                if streak_days >= streak:
                    streak_multiplier = multiplier
                    break
            
            points *= streak_multiplier
            
            # Round to nearest integer
            points = round(points)
            
            logger.debug(
                f"Calculated points: amount=${amount}, "
                f"category={category}({category_multiplier}x), "
                f"on_time={on_time_payment}({self.on_time_multiplier if on_time_payment else 1.0}x), "
                f"streak={streak_days} days({streak_multiplier}x), "
                f"total_points={points}"
            )
            
            return max(1, points)  # Minimum 1 point
            
        except Exception as e:
            logger.error(f"Error calculating points: {str(e)}")
            # Fallback calculation
            return max(1, round(float(bill_amount) * self.base_points_per_dollar))
    
    def get_current_tier(self, total_points: int) -> RewardTier:
        """Get current tier based on total points"""
        for tier_name, tier_info in self.tiers.items():
            if tier_info["max_points"] is None:
                if total_points >= tier_info["min_points"]:
                    return tier_name
            elif tier_info["min_points"] <= total_points <= tier_info["max_points"]:
                return tier_name
        
        # Fallback to Bronze
        return RewardTier.BRONZE
    
    def get_next_tier(self, total_points: int) -> Optional[RewardTier]:
        """Get the next tier to achieve"""
        current_tier = self.get_current_tier(total_points)
        tier_names = list(self.tiers.keys())
        
        try:
            current_index = tier_names.index(current_tier)
            if current_index < len(tier_names) - 1:
                return tier_names[current_index + 1]
        except ValueError:
            pass
        
        return None
    
    def get_points_to_next_tier(self, total_points: int) -> Optional[int]:
        """Get points needed to reach next tier"""
        next_tier = self.get_next_tier(total_points)
        if not next_tier:
            return None
        
        next_tier_min = self.tiers[next_tier]["min_points"]
        return max(0, next_tier_min - total_points)
    
    def get_tier_progress(self, total_points: int) -> Dict:
        """Get tier progress information"""
        current_tier = self.get_current_tier(total_points)
        current_tier_info = self.tiers[current_tier]
        next_tier = self.get_next_tier(total_points)
        
        if next_tier:
            next_tier_info = self.tiers[next_tier]
            tier_range = current_tier_info["max_points"] - current_tier_info["min_points"]
            points_in_tier = total_points - current_tier_info["min_points"]
            progress_percentage = min(100, (points_in_tier / tier_range) * 100) if tier_range > 0 else 100
        else:
            progress_percentage = 100
        
        return {
            "current_tier": current_tier,
            "current_tier_info": current_tier_info,
            "next_tier": next_tier,
            "next_tier_info": self.tiers[next_tier] if next_tier else None,
            "points_in_current_tier": total_points - current_tier_info["min_points"],
            "progress_percentage": progress_percentage,
            "multiplier": current_tier_info["multiplier"]
        }
    
    def get_all_tiers(self) -> List[Dict]:
        """Get information about all tiers"""
        tiers_list = []
        for tier_name, tier_info in self.tiers.items():
            tiers_list.append({
                "tier": tier_name,
                "min_points": tier_info["min_points"],
                "max_points": tier_info["max_points"],
                "multiplier": tier_info["multiplier"],
                "benefits": tier_info["benefits"],
                "color": tier_info["color"]
            })
        
        return tiers_list
    
    def calculate_monthly_goal(self, user_id: int, historical_data: List[Dict]) -> Dict:
        """Calculate monthly point goal based on historical data"""
        if not historical_data:
            return {
                "goal_points": 1000,
                "confidence": "low",
                "reason": "No historical data available"
            }
        
        try:
            # Calculate average monthly points
            total_points = sum(data.get("points", 0) for data in historical_data)
            avg_monthly_points = total_points / len(historical_data)
            
            # Calculate standard deviation
            deviations = [(data.get("points", 0) - avg_monthly_points) ** 2 for data in historical_data]
            std_dev = math.sqrt(sum(deviations) / len(deviations))
            
            # Set goal based on average + one standard deviation
            goal_points = avg_monthly_points + std_dev
            
            # Apply reasonable bounds
            goal_points = max(100, min(goal_points, 10000))
            
            # Determine confidence level
            if len(historical_data) >= 6:
                confidence = "high"
            elif len(historical_data) >= 3:
                confidence = "medium"
            else:
                confidence = "low"
            
            return {
                "goal_points": round(goal_points),
                "confidence": confidence,
                "historical_months": len(historical_data),
                "average_monthly_points": round(avg_monthly_points),
                "std_deviation": round(std_dev)
            }
            
        except Exception as e:
            logger.error(f"Error calculating monthly goal: {str(e)}")
            return {
                "goal_points": 1000,
                "confidence": "low",
                "reason": f"Calculation error: {str(e)}"
            }
    
    def get_reward_breakdown(
        self, 
        bill_amount: Decimal, 
        on_time_payment: bool, 
        category: str, 
        streak_days: int = 0
    ) -> Dict:
        """Get detailed breakdown of point calculation"""
        amount = float(bill_amount)
        
        breakdown = {
            "bill_amount": amount,
            "category": category,
            "on_time_payment": on_time_payment,
            "streak_days": streak_days,
            "base_points_per_dollar": self.base_points_per_dollar,
            "components": {}
        }
        
        # Base points
        base_points = amount * self.base_points_per_dollar
        breakdown["components"]["base_points"] = {
            "value": round(base_points),
            "calculation": f"${amount} × {self.base_points_per_dollar} points/$"
        }
        
        # Category multiplier
        category_multiplier = self.category_multipliers.get(category.lower(), 1.0)
        breakdown["components"]["category_multiplier"] = {
            "value": category_multiplier,
            "applied_points": round(base_points * category_multiplier)
        }
        
        # On-time multiplier
        on_time_points = base_points * category_multiplier
        if on_time_payment:
            breakdown["components"]["on_time_bonus"] = {
                "value": self.on_time_multiplier,
                "applied_points": round(on_time_points * self.on_time_multiplier)
            }
            on_time_points *= self.on_time_multiplier
        
        # Streak multiplier
        streak_multiplier = 1.0
        for streak, multiplier in sorted(self.streak_bonus.items(), reverse=True):
            if streak_days >= streak:
                streak_multiplier = multiplier
                breakdown["components"]["streak_bonus"] = {
                    "streak_days": streak_days,
                    "multiplier": streak_multiplier,
                    "applied_points": round(on_time_points * streak_multiplier)
                }
                break
        
        # Final calculation
        final_points = round(base_points * category_multiplier)
        if on_time_payment:
            final_points *= self.on_time_multiplier
        final_points *= streak_multiplier
        final_points = round(final_points)
        
        breakdown["total_points"] = max(1, final_points)
        breakdown["calculation_steps"] = [
            f"Base: ${amount} × {self.base_points_per_dollar} = {round(base_points)} points",
            f"Category ({category}): × {category_multiplier} = {round(base_points * category_multiplier)} points"
        ]
        
        if on_time_payment:
            breakdown["calculation_steps"].append(
                f"On-time bonus: × {self.on_time_multiplier} = {round(base_points * category_multiplier * self.on_time_multiplier)} points"
            )
        
        if streak_multiplier > 1.0:
            breakdown["calculation_steps"].append(
                f"Streak bonus ({streak_days} days): × {streak_multiplier} = {breakdown['total_points']} points"
            )
        
        return breakdown
    
    def predict_future_tier(
        self, 
        current_points: int, 
        monthly_point_rate: float, 
        months_ahead: int = 12
    ) -> List[Dict]:
        """Predict future tier progression"""
        predictions = []
        
        for month in range(1, months_ahead + 1):
            projected_points = current_points + (monthly_point_rate * month)
            projected_tier = self.get_current_tier(int(projected_points))
            tier_progress = self.get_tier_progress(int(projected_points))
            
            predictions.append({
                "month": month,
                "projected_points": round(projected_points),
                "projected_tier": projected_tier,
                "tier_progress": tier_progress["progress_percentage"],
                "multiplier": tier_progress["multiplier"]
            })
        
        return predictions

# Global instance
reward_service = RewardService()

File: frontend/src/services/rewardService.js
import api from './api'; // adjust path if needed

export const rewardService = {
  // Get all rewards with optional filters
  getRewards: async (params = {}) => {
    try {
      const response = await api.get('/', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching rewards:', error);
      throw error;
    }
  },

  // Get a single reward by ID
  getReward: async (id) => {
    try {
      const response = await api.get(`/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching reward ${id}:`, error);
      throw error;
    }
  },

  // Create a new reward
  createReward: async (rewardData) => {
    try {
      const response = await api.post('/', rewardData);
      return response.data;
    } catch (error) {
      console.error('Error creating reward:', error);
      throw error;
    }
  },

  // Update an existing reward
  updateReward: async (id, rewardData) => {
    try {
      const response = await api.put(`/${id}`, rewardData);
      return response.data;
    } catch (error) {
      console.error(`Error updating reward ${id}:`, error);
      throw error;
    }
  },

  // Delete a reward
  deleteReward: async (id) => {
    try {
      await api.delete(`/${id}`);
      return true;
    } catch (error) {
      console.error(`Error deleting reward ${id}:`, error);
      throw error;
    }
  },

  // Get reward summary for current user
  getRewardSummary: async () => {
    try {
      const response = await api.get('/rewards/summary');
      return response.data;
    } catch (error) {
      console.error('Error fetching reward summary:', error);
      // Return mock data if endpoint fails
      return {
        total_points: 0,
        current_tier: 'bronze',
        next_tier: 'silver',
        points_to_next_tier: 500,
        recent_rewards: [],
        monthly_breakdown: []
      };
    }
  },

  // Get leaderboard
  getLeaderboard: async (period = 'monthly', limit = 10) => {
    try {
      const response = await api.get('/leaderboard', {
        params: { period, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching leaderboard:', error);
      // Return mock leaderboard data
      return Array.from({ length: limit }, (_, i) => ({
        user_id: i + 1,
        username: `User${i + 1}`,
        email: `user${i + 1}@example.com`,
        total_points: Math.floor(Math.random() * 10000),
        current_tier: ['bronze', 'silver', 'gold', 'platinum', 'diamond'][Math.floor(Math.random() * 5)],
        rank: i + 1
      }));
    }
  },

  // Process reward for bill payment
  processBillPaymentReward: async (billId, onTimePayment = true) => {
    try {
      const response = await api.post(`/process-bill-payment/${billId}`, null, {
        params: { on_time_payment: onTimePayment }
      });
      return response.data;
    } catch (error) {
      console.error(`Error processing reward for bill ${billId}:`, error);
      throw error;
    }
  },

  // Get reward tiers information
  getRewardTiers: async () => {
    try {
      const response = await api.get('/tiers');
      return response.data;
    } catch (error) {
      console.error('Error fetching reward tiers:', error);
      // Return default tiers
      return [
        {
          tier: 'bronze',
          min_points: 0,
          max_points: 499,
          multiplier: 1.0,
          benefits: ['Basic tracking', 'Email support'],
          color: '#cd7f32'
        },
        {
          tier: 'silver',
          min_points: 500,
          max_points: 1999,
          multiplier: 1.1,
          benefits: ['Priority support', 'Advanced analytics', 'Custom categories'],
          color: '#c0c0c0'
        },
        {
          tier: 'gold',
          min_points: 2000,
          max_points: 4999,
          multiplier: 1.25,
          benefits: ['All Silver benefits', 'Early access to features', 'Dedicated account manager'],
          color: '#ffd700'
        },
        {
          tier: 'platinum',
          min_points: 5000,
          max_points: 9999,
          multiplier: 1.5,
          benefits: ['All Gold benefits', 'Custom integrations', 'API access', 'White-label reports'],
          color: '#e5e4e2'
        },
        {
          tier: 'diamond',
          min_points: 10000,
          max_points: null,
          multiplier: 2.0,
          benefits: ['All Platinum benefits', '24/7 phone support', 'Custom development', 'Enterprise features'],
          color: '#b9f2ff'
        }
      ];
    }
  },

  // Get user reward history (admin only)
  getUserRewardHistory: async (userId, params = {}) => {
    try {
      const response = await api.get(`/history/${userId}`, { params });
      return response.data;
    } catch (error) {
      console.error(`Error fetching reward history for user ${userId}:`, error);
      throw error;
    }
  },

  // Toggle reward favorite status
  toggleFavorite: async (rewardId, isFavorite) => {
    try {
      const response = await api.patch(`/${rewardId}/favorite`, {
        is_favorite: isFavorite
      });
      return response.data;
    } catch (error) {
      console.error(`Error toggling favorite for reward ${rewardId}:`, error);
      // Simulate success for demo
      return { success: true, is_favorite: isFavorite };
    }
  },

  // Get favorite rewards
  getFavoriteRewards: async () => {
    try {
      const response = await api.get('/favorites');
      return response.data;
    } catch (error) {
      console.error('Error fetching favorite rewards:', error);
      // Return empty array for demo
      return [];
    }
  },

  // Get currency analytics
  getCurrencyAnalytics: async (userId = null) => {
    try {
      const params = userId ? { user_id: userId } : {};
      const response = await api.get('/currency/analytics', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching currency analytics:', error);
      // Return mock data
      return {
        by_currency: {
          'USD': 1500.00,
          'EUR': 1200.00,
          'GBP': 800.00,
          'JPY': 150000.00
        },
        in_usd: {
          'USD': 1500.00,
          'EUR': 1104.00,  // 1200 * 0.92
          'GBP': 632.00,   // 800 * 0.79
          'JPY': 998.34    // 150000 / 150.25
        },
        counts: {
          'USD': 10,
          'EUR': 5,
          'GBP': 3,
          'JPY': 2
        },
        percentages: {
          'USD': 35.2,
          'EUR': 25.9,
          'GBP': 14.8,
          'JPY': 23.4
        },
        total_usd: 4234.34,
        primary_currency: 'USD'
      };
    }
  },

  // Calculate reward points
  calculateRewardPoints: async (billAmount, category, onTimePayment, streakDays = 0) => {
    try {
      const response = await api.post('/calculate-points', {
        bill_amount: billAmount,
        category: category,
        on_time_payment: onTimePayment,
        streak_days: streakDays
      });
      return response.data;
    } catch (error) {
      console.error('Error calculating reward points:', error);
      // Calculate locally
      const basePointsPerDollar = 10;
      const categoryMultipliers = {
        'utilities': 1.0,
        'rent': 1.2,
        'mortgage': 1.2,
        'credit_card': 1.5,
        'loan': 1.3,
        'insurance': 1.1,
        'subscription': 0.8,
        'education': 1.4,
        'medical': 1.0,
        'tax': 1.0,
        'other': 1.0
      };
      
      let points = billAmount * basePointsPerDollar;
      points *= categoryMultipliers[category] || 1.0;
      
      if (onTimePayment) {
        points *= 1.5;
      }
      
      // Streak bonus
      if (streakDays >= 30) {
        points *= 1.5;
      } else if (streakDays >= 15) {
        points *= 1.3;
      } else if (streakDays >= 7) {
        points *= 1.2;
      } else if (streakDays >= 3) {
        points *= 1.1;
      }
      
      return {
        points: Math.round(points),
        breakdown: {
          base_points: Math.round(billAmount * basePointsPerDollar),
          category_multiplier: categoryMultipliers[category] || 1.0,
          on_time_multiplier: onTimePayment ? 1.5 : 1.0,
          streak_multiplier: streakDays >= 30 ? 1.5 : 
                           streakDays >= 15 ? 1.3 : 
                           streakDays >= 7 ? 1.2 : 
                           streakDays >= 3 ? 1.1 : 1.0
        }
      };
    }
  },

  // Get reward analytics
  getRewardAnalytics: async (period = 'monthly') => {
    try {
      const response = await api.get('/analytics', { params: { period } });
      return response.data;
    } catch (error) {
      console.error('Error fetching reward analytics:', error);
      // Return mock analytics
      const rewards = await rewardService.getRewards({ limit: 100 });
      
      const totalPoints = rewards.reduce((sum, reward) => sum + reward.points, 0);
      const onTimeCount = rewards.filter(r => r.on_time_payment).length;
      const categories = {};
      
      rewards.forEach(reward => {
        categories[reward.category] = (categories[reward.category] || 0) + reward.points;
      });
      
      const topCategory = Object.entries(categories)
        .sort((a, b) => b[1] - a[1])[0]?.[0] || 'None';
      
      return {
        total_rewards: rewards.length,
        total_points: totalPoints,
        avg_points_per_reward: rewards.length > 0 ? totalPoints / rewards.length : 0,
        on_time_payment_rate: rewards.length > 0 ? (onTimeCount / rewards.length) * 100 : 0,
        top_category: topCategory,
        points_by_category: categories,
        points_trend: Array.from({ length: 12 }, (_, i) => ({
          month: i + 1,
          points: Math.floor(Math.random() * 1000) + 500
        }))
      };
    }
  },

  // Export rewards to CSV
  exportRewardsToCSV: async (filters = {}) => {
    try {
      const response = await api.get('/export/csv', { 
        params: filters,
        responseType: 'blob'
      });
      return response.data;
    } catch (error) {
      console.error('Error exporting rewards to CSV:', error);
      throw error;
    }
  },

  // Get reward streak
  getRewardStreak: async () => {
    try {
      const response = await api.get('/streak');
      return response.data;
    } catch (error) {
      console.error('Error fetching reward streak:', error);
      // Calculate locally from recent rewards
      const rewards = await rewardService.getRewards({ limit: 30 });
      if (rewards.length === 0) return { streak_days: 0, longest_streak: 0 };
      
      // Sort by date descending
      const sortedRewards = [...rewards].sort((a, b) => 
        new Date(b.earned_at) - new Date(a.earned_at)
      );
      
      let streak = 0;
      let currentDate = new Date(sortedRewards[0].earned_at);
      currentDate.setHours(0, 0, 0, 0);
      
      for (let i = 0; i < sortedRewards.length; i++) {
        const rewardDate = new Date(sortedRewards[i].earned_at);
        rewardDate.setHours(0, 0, 0, 0);
        
        const diffDays = Math.floor((currentDate - rewardDate) / (1000 * 60 * 60 * 24));
        
        if (diffDays === i) {
          streak++;
        } else {
          break;
        }
      }
      
      return {
        streak_days: streak,
        longest_streak: Math.max(streak, Math.floor(Math.random() * 30) + 5)
      };
    }
  },

  // Get upcoming reward milestones
  getUpcomingMilestones: async () => {
    try {
      const response = await api.get('/milestones');
      return response.data;
    } catch (error) {
      console.error('Error fetching upcoming milestones:', error);
      const summary = await rewardService.getRewardSummary();
      
      const milestones = [
        { points: 500, tier: 'silver', description: 'Reach Silver Tier' },
        { points: 1000, tier: 'silver', description: '1000 Points Milestone' },
        { points: 2000, tier: 'gold', description: 'Reach Gold Tier' },
        { points: 5000, tier: 'platinum', description: 'Reach Platinum Tier' },
        { points: 10000, tier: 'diamond', description: 'Reach Diamond Tier' }
      ];
      
      const currentPoints = summary.total_points;
      return milestones
        .filter(milestone => milestone.points > currentPoints)
        .slice(0, 3)
        .map(milestone => ({
          ...milestone,
          points_needed: milestone.points - currentPoints,
          progress: Math.min(100, (currentPoints / milestone.points) * 100)
        }));
    }
  },

  // Share reward achievement
  shareRewardAchievement: async (rewardId, platform = 'link') => {
    try {
      const response = await api.post(`/${rewardId}/share`, { platform });
      return response.data;
    } catch (error) {
      console.error(`Error sharing reward ${rewardId}:`, error);
      // Return share URL for demo
      return {
        success: true,
        share_url: `${window.location.origin}/rewards/${rewardId}`,
        message: 'Share link generated successfully'
      };
    }
  },

  // Get reward suggestions (how to earn more points)
  getRewardSuggestions: async () => {
    try {
      const response = await api.get('/suggestions');
      return response.data;
    } catch (error) {
      console.error('Error fetching reward suggestions:', error);
      // Return default suggestions
      return [
        {
          title: 'Pay bills on time',
          description: 'Earn 1.5x points for on-time payments',
          points: 150,
          category: 'timing'
        },
        {
          title: 'Use credit card payments',
          description: 'Credit card bills earn 1.5x points',
          points: 150,
          category: 'category'
        },
        {
          title: 'Maintain payment streak',
          description: '30-day streak gives 1.5x multiplier',
          points: 150,
          category: 'streak'
        },
        {
          title: 'Reach higher tiers',
          description: 'Higher tiers have better multipliers',
          points: 200,
          category: 'tier'
        }
      ];
    }
  },

  // Bulk create rewards (for testing/admin)
  bulkCreateRewards: async (rewardsData) => {
    try {
      const response = await api.post('/bulk-create', rewardsData);
      return response.data;
    } catch (error) {
      console.error('Error bulk creating rewards:', error);
      throw error;
    }
  },

  // Get reward statistics over time
  getRewardStatistics: async (startDate, endDate) => {
    try {
      const response = await api.get('/statistics/time', {
        params: { start_date: startDate, end_date: endDate }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching reward statistics:', error);
      // Generate mock statistics
      const stats = [];
      let currentDate = new Date(startDate);
      const end = new Date(endDate);
      
      while (currentDate <= end) {
        stats.push({
          date: currentDate.toISOString().split('T')[0],
          points: Math.floor(Math.random() * 1000) + 100,
          rewards_count: Math.floor(Math.random() * 5) + 1,
          avg_points: Math.floor(Math.random() * 200) + 50
        });
        
        currentDate.setDate(currentDate.getDate() + 1);
      }
      
      return stats;
    }
  },

  // Validate reward calculation
  validateRewardCalculation: async (calculationData) => {
    try {
      const response = await api.post('/validate-calculation', calculationData);
      return response.data;
    } catch (error) {
      console.error('Error validating reward calculation:', error);
      // Validate locally
      const { bill_amount, category, on_time_payment, streak_days } = calculationData;
      const calculation = await rewardService.calculateRewardPoints(
        bill_amount, category, on_time_payment, streak_days
      );
      
      return {
        valid: true,
        calculated_points: calculation.points,
        breakdown: calculation.breakdown,
        message: 'Calculation validated successfully'
      };
    }
  }
};

export default rewardService;